FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt

```diff
@@ -256,6 +268,324 @@ object PgpMsg {
+  // invoke in the test as:
+  // val session = Session.getInstance(Properties())
+  // val msg = MimeMessage(session, stream)
+  fun decodeMimeMessage(msg: MimeMessage): MimeContent {
+    var signature: String? = null
+    var html: StringBuilder? = null
+    var text: StringBuilder? = null
+    val attachments = mutableListOf<BodyPart>()
+
+    val stack = ArrayDeque<Part>()
+    stack.push(msg)
+
+    while (stack.isNotEmpty()) {
+      val part = stack.pop()
+      if (part.isMimeType(JavaEmailConstants.MIME_TYPE_MULTIPART)) {
+        // multi-part, break down into separate parts
+        val multipart = part.content as Multipart
+        val list = mutableListOf<Part>()
+        val n = multipart.count
+        for (i in 0 until n) {
+          list.add(multipart.getBodyPart(i))
+        }
+        for (i in 0 until n) {
+          stack.push(list[n - i - 1])
+        }
+      } else {
+        // single part, analyze content type and extract some data
+        part as BodyPart
+        // println("parse: '${part.contentType}' $contentType '${part.fileName}' ${part.size}")
+        when (part.contentType.split(';').first().trim()) {
+          "application/pgp-signature" -> {
+            signature = String(part.inputStream.readBytes(), StandardCharsets.US_ASCII)
+          }
+
+          // this one was not in the Typescript, but I had to add it to pass some tests
+          "message/rfc822" -> {
+            stack.push(part.content as Part)
+          }
+
+          "text/html" -> {
+            if (!part.hasFileName()) {
+              if (html == null) html = StringBuilder()
+              html.append(part.content)
+            }
+          }
+
+          "text/plain" -> {
+            if (!part.hasFileName() || part.isInline()) {
+              if (text == null) {
+                text = StringBuilder()
+              } else {
+                text.append("\n\n")
+              }
+              text.append(part.content)
+            }
+          }
+
+          "text/rfc822-headers" -> {} // skip
+
+          else -> {
+            attachments.add(part)
+          }
+        }
+      }
+    }
+    return MimeContent(
+      attachments = attachments,
+      signature = signature,
+      subject = msg.subject,
+      html = html?.toString(),
+      text = text?.toString(),
+      from = msg.from,
+      to = msg.getRecipients(Message.RecipientType.TO),
+      cc = msg.getRecipients(Message.RecipientType.CC),
+      bcc = msg.getRecipients(Message.RecipientType.BCC)
+    )
+  }
+
+  // Typescript:  public static processDecoded = (decoded: MimeContent): MimeProcessedMsg
+  fun processDecodedMimeMessage(decoded: MimeContent): MimeProcessedMsg {
+    val blocks = mutableListOf<MsgBlock>()
+    if (decoded.text != null) {
+      val blocksFromTextPart = MsgBlockParser.detectBlocks(decoded.text)
+      val suitableBlock = blocksFromTextPart.firstOrNull {
+        it.type == MsgBlock.Type.ENCRYPTED_MSG || it.type == MsgBlock.Type.SIGNED_MSG
```

Comment:
Can it be simplified with `it.type in MsgBlock.Type.wellKnownBlockTypes`?

Reply:
Yes, I think so, didn't notice this.